plugins/obs-qsv11/obs-qsv265.c

```diff
@@ -0,0 +1,1118 @@
+
+#define warn(format, ...) do_log(LOG_WARNING, format, ##__VA_ARGS__)
+#define info(format, ...) do_log(LOG_INFO, format, ##__VA_ARGS__)
+#define debug(format, ...) do_log(LOG_DEBUG, format, ##__VA_ARGS__)
+
+static uint8_t *ff_find_startcode_internal(const uint8_t *p, const uint8_t *end)
+{
+	const uint8_t *a = p + 4 - ((intptr_t)p & 3);
+
+	for (end -= 3; p < a && p < end; p++) {
+		if (p[0] == 0 && p[1] == 0 && p[2] == 1)
+			return (uint8_t *)p;
+	}
+
+	for (end -= 3; p < end; p += 4) {
+		uint32_t x = *(const uint32_t *)p;
+
+		if ((x - 0x01010101) & (~x) & 0x80808080) {
+			if (p[1] == 0) {
+				if (p[0] == 0 && p[2] == 1)
+					return (uint8_t *)p;
+				if (p[2] == 0 && p[3] == 1)
+					return (uint8_t *)(p + 1);
+			}
+
+			if (p[3] == 0) {
+				if (p[2] == 0 && p[4] == 1)
+					return (uint8_t *)(p + 2);
+				if (p[4] == 0 && p[5] == 1)
+					return (uint8_t *)(p + 3);
+			}
+		}
+	}
+
+	for (end += 3; p < end; p++) {
+		if (p[0] == 0 && p[1] == 0 && p[2] == 1)
+			return (uint8_t *)p;
+	}
+
+	return (uint8_t *)(end + 3);
+}
+
+static uint8_t *obs_find_annexB_startcode(const uint8_t *p, const uint8_t *end)
+{
+	uint8_t *out = ff_find_startcode_internal(p, end);
+	if (p < out && out < end && !out[-1])
+		out--;
+	return out;
+}
+
+/* ------------------------------------------------------------------------- */
+
+struct obs_qsv {
+	obs_encoder_t *encoder;
+
+	qsv_param_t params;
+	qsv_t *context;
+
+	DARRAY(uint8_t) packet_data;
+
+	uint8_t *extra_data;
+	uint8_t *sei;
+
+	size_t extra_data_size;
+	size_t sei_size;
+
+	os_performance_token_t *performance_token;
+};
+
+/* ------------------------------------------------------------------------- */
+
+static CRITICAL_SECTION g_QsvCs;
+static unsigned short g_verMajor;
+static unsigned short g_verMinor;
+static int64_t g_pts2dtsShift;
+static int64_t g_prevDts;
+static bool g_bFirst;
+
+static const char *obs_qsv_hevc_getname(void *type_data)
+{
+	UNUSED_PARAMETER(type_data);
+	return "QuickSync HEVC Encoder";
+}
+
+static void obs_qsv_hevc_stop(void *data);
+
+static void clear_data(struct obs_qsv *obsqsv)
+{
+	if (obsqsv->context) {
+		EnterCriticalSection(&g_QsvCs);
+		qsv_hevc_encoder_close(obsqsv->context);
+		obsqsv->context = NULL;
+		LeaveCriticalSection(&g_QsvCs);
+
+		bfree(obsqsv->sei);
+		obsqsv->sei = NULL;
+
+		bfree(obsqsv->extra_data);
+		obsqsv->extra_data = NULL;
+	}
+}
+
+static void obs_qsv_hevc_destroy(void *data)
+{
+	struct obs_qsv *obsqsv = (struct obs_qsv *)data;
+
+	if (obsqsv) {
+		os_end_high_performance(obsqsv->performance_token);
+		clear_data(obsqsv);
+		da_free(obsqsv->packet_data);
+		bfree(obsqsv);
+	}
+}
+
+static void obs_qsv_hevc_defaults(obs_data_t *settings)
+{
+	obs_data_set_default_string(settings, "target_usage", "veryfast");
+	obs_data_set_default_int(settings, "bitrate", 2500);
+	obs_data_set_default_int(settings, "max_bitrate", 30000);
+	obs_data_set_default_string(settings, "profile", "Main 10");
```

Comment:
Default to "main" but dynamically choose MFX_PROFILE_HEVC_MAIN10 if the bit-depth is 10. (Same logic that NVENC uses.)

Reply:
Automatically switches to main10 if P010 is active now.